In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Import Data

In [2]:
# To use the os.path.join function, you first have to define the path
path = r'/Users/oluwaseunodusanya/Desktop/06.2023_Instacart Basket Analysis'

In [3]:
#Import Product Data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

In [4]:
#Import orders_wrangled Data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col=False)

# Consistency Check for df_prods

In [5]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

In [6]:
# To test for mixed data
df_test = pd.DataFrame()

In [7]:
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test

mix
0     a
1     b
2     1
3  True

In [9]:
# To find the number of missing data per column
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
# To create a subset of missing data from product_name
df_nan =  df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
#To find the number of rows and columns in df_prods
df_prods.shape 

(49693, 5)

In [13]:
# To highlight non-missing values in your new dataframe as opposed to missing values
df_prods_clean =  df_prods[df_prods['product_name'].isnull() == False]

In [14]:
df_prods_clean.shape

(49677, 5)

In [17]:
# Another method to drop missing value
# From a particular column 
df_prods.dropna(subset = ['product_name'], inplace = True)
# From the entire product data
df_prods.dropna(inplace = True)

In [18]:
# To remove duplicates from new df_prods_clean, create a suset to highlight the duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
df_prods_clean.shape

(49677, 5)

In [21]:
# Remove duplicates by using the drop fuction
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

In [23]:
# To export df_prods after consistency check
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5 Consistency Check for df_ords

In [24]:
# Step 2 Run the df.describe() function on your df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

## whether anything about the data looks off or should be investigated further
Looking at the order_number column, It has a minimum of '1' and maximum of 100. But the 25%, 50% and 75% looks off and nee dto be investigated further

In [25]:
# To drop the Unnamed column for df_ords
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [26]:
# Step 3 Check for mixed-type data
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


## No mixed-type data found

In [27]:
# Step 5 Run a check for missing values
df_ords.isnull().sum()

order_id                      0
user_id                       0
eval_set                      0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [28]:
# To create a subset of the missing values in question in order to view them
df_ords_nan =  df_ords[df_ords['days_since_last_order'].isnull() == True]

In [29]:
df_ords_nan

order_id  user_id eval_set  order_number  orders_day_of_week  \
0         2539329        1    prior             1                   2   
11        2168274        2    prior             1                   2   
26        1374495        3    prior             1                   1   
39        3343014        4    prior             1                   6   
45        2717275        5    prior             1                   3   
...           ...      ...      ...           ...                 ...   
3420930    969311   206205    prior             1                   4   
3420934   3189322   206206    prior             1                   3   
3421002   2166133   206207    prior             1                   6   
3421019   2227043   206208    prior             1                   1   
3421069   3154581   206209    prior             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

## report your findings and propose an explanation for any missing values you find
column days_since_last_order contains '206209' missing data. The reason for the missing value narrows down to the majority of customers that had their first order. Meaning, there would not be any record for the day since last order for these first order. 

In [ ]:
# Step 6 To Address the missing values using an appropriate method
I believe making any modifications to these missing values would not be necessary as they have meaning in themselves. Customers with order number of 1 can represent new customers, without day_since_last_order values

In [30]:
# Step 7 check for duplicate values
df_ords_dups = df_ords[df_ords.duplicated()]

In [31]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

## No duplicates found

In [ ]:
# Step 8 Address the duplicates using an appropriate method
No duplicates found

# Step 9 Export your final, cleaned df_prods and df_ords data as “.csv”

In [ ]:
# To export df_ords after consistency check
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))